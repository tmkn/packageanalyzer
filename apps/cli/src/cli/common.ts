import { Command } from "clipanion";

import { AbstractReport } from "../../../../packages/shared/src/reports/Report.js";
import { ReportService } from "../../../../packages/shared/src/reports/ReportService.js";
import { Formatter, type IFormatter } from "../../../../packages/shared/src/utils/formatter.js";
import { NodeHost } from "../../../../packages/node/src/host/NodeHost.js";

export abstract class CliCommand<
    T extends AbstractReport<any> | AbstractReport<any>[]
> extends Command {
    abstract getReports(): T | Promise<T>;

    exitCode = 0;

    beforeProcess: ((report: T) => void) | undefined = undefined;

    async execute(): Promise<number | void> {
        const host = new NodeHost(this.context.stdout, this.context.stderr);

        try {
            const reports = await this.getReports();
            const reportService = new ReportService(
                {
                    reports: Array.isArray(reports) ? reports : [reports]
                },
                host
            );

            this.beforeProcess?.(reports);
            this.exitCode = (await reportService.process()) ?? 0;
        } catch (e: unknown) {
            const stderrFormatter: IFormatter = new Formatter(host.getStderrWriter());

            if (e) stderrFormatter.writeLine(e?.toString());
            this.exitCode = 1;
        } finally {
            await Promise.all([host.getStdoutWriter().flush(), host.getStderrWriter().flush()]);
        }

        return this.exitCode;
    }
}
