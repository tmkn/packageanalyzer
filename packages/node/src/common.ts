import * as path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";

import dayjs from "dayjs";

import type { DependencyTypes } from "../../shared/src/reports/Validation.js";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const __dirname = path.dirname(fileURLToPath(import.meta.url));
        // adjust when moving this function to another folder
        const file = path.join(__dirname, "..", "..", "..", "package.json");
        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}
