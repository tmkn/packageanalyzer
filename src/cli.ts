#!/usr/bin/env node

const pkg = require("./../../package.json");

import * as dayjs from "dayjs";

import { resolveFromFolder } from "./resolvers/folderResolver";
import { resolveFromName } from "./resolvers/nameResolver";
import { npmOnline } from "./providers/onlineProvider";
import {
    PackageAnalytics,
    LicenseSummary,
    VersionSummary,
    getNameAndVersion,
    groupPackagesByLicense
} from "./analyzer";

let commandFound = false;

process.argv.forEach((arg, i) => {
    if (arg === "-v" && !commandFound) {
        console.log(`Version ${pkg.version}`);
        commandFound = true;
    } else if (arg === "-h" && !commandFound) {
        showHelp();
        commandFound = true;
    } else if (arg === "-f" && !commandFound) {
        let folder = process.argv[i + 1];

        cliResolveFolder(folder);
        commandFound = true;
    } else if (arg === "-o" && !commandFound) {
        let name = process.argv[i + 1];

        cliResolveName(name);
        commandFound = true;
    }
});

if (!commandFound) {
    console.log(`No command found`);
    showHelp();
}

function showHelp(): void {
    console.log(`Package Analyzer ${pkg.version}`);
    console.log(`Options:`);
    console.log(
        `${`-f`.padEnd(10)} ${`Analyze a local folder`.padEnd(60)} e.g. "npa -f path/to/project"`
    );
    console.log(
        `${`-o`.padEnd(10)} ${`Analyze a package, version optional, default latest`.padEnd(
            60
        )} e.g. "npa -o typescript(@3.5.1)"`
    );
    console.log(`${`-v`.padEnd(10)} ${`Prints version`.padEnd(60)}`);
    console.log(`${`-h`.padEnd(10)} ${`Display help`.padEnd(60)}`);
}

async function cliResolveFolder(folder: string | undefined): Promise<void> {
    if (typeof folder === "undefined") {
        console.log(`Missing folder path\n`);
        showHelp();

        return;
    }

    try {
        let pa: PackageAnalytics = await resolveFromFolder(folder);

        printStatistics(pa);
    } catch (e) {
        //console.log(e);
    }
}

async function cliResolveName(pkgName: string | undefined): Promise<void> {
    if (typeof pkgName === "undefined") {
        console.log(`Missing package name\n`);
        showHelp();

        return;
    }

    try {
        const [name, version] = getNameAndVersion(pkgName);
        let pa: PackageAnalytics;

        if (typeof version === "undefined") pa = await resolveFromName(name, npmOnline);
        else pa = await resolveFromName([name, version], npmOnline);

        printStatistics(pa);
    } catch (e) {
        console.log(e);
    }
}

function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(`Statistics for ${pa.fullName}\n`);

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa.loops, padding, paddingLeft);
    printLicenseInfo(pa.licenses, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
    }
}

function daysAgo(date: number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(loops: PackageAnalytics[], padding: number, paddingLeft: number): void {
    console.log(`${`Loops:`.padEnd(padding)}${loops.length}`);

    if (loops.length > 0) {
        let [first] = loops.map(l => l.pathString).sort();

        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(allLicenses: LicenseSummary, paddingLeft: number): void {
    let summary = new Map<string, number>();
    let longestLine = 0;

    for (const [, licenses] of allLicenses) {
        for (const [, license] of licenses) {
            let specificLicense = summary.get(license);

            if (!specificLicense) {
                summary.set(license, 1);
            } else {
                summary.set(license, specificLicense + 1);
            }

            if (license.length > longestLine) longestLine = license.length;
        }
    }

    console.log(`Licenses:`);
    for (const { license, names } of groupPackagesByLicense(allLicenses)) {
        let threshold = 4;
        let samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${names.length -
                    threshold} more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }

    /*let longestLine = 0;

    for(const [name, licenses] of allLicenses) {
        for(const [version, license] of licenses) {
            let fullName = `${name}@${version}`;

            if(fullName.length > longestLine)
                longestLine = fullName.length;
        }
    }

    for(const [name, licenses] of allLicenses) {
        for(const [version, license] of licenses) {
            console.log(`${`${name}@${version}`.padEnd(longestLine + 1)}${license}`.padStart(8));
        }
    }*/
}
