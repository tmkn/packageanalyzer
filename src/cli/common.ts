import * as path from "path";
import * as fs from "fs";

import * as dayjs from "dayjs";
import * as chalk from "chalk";

import { PackageAnalytics, VersionSummary, GroupedLicenseSummary } from "../analyzers/package";
import { DependencyTypes } from "../visitors/visitor";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const file = path.join(__dirname, "./../../package.json");

        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch (e) {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

export function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(chalk.bold(`Statistics for ${pa.fullName}\n`));

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa, padding, paddingLeft);
    printLicenseInfo(pa.licensesByGroup, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest package:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
        console.log(`${`Newest package path:`.padEnd(padding)}${newest.pathString}`);
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest package:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
        console.log(`${`Oldest package path:`.padEnd(padding)}${oldest.pathString}`);
    }
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(pa: PackageAnalytics, padding: number, paddingLeft: number): void {
    const { loops, loopPathMap, distinctLoopCount } = pa;

    console.log(`${`Loops:`.padEnd(padding)}${loops.length} (${distinctLoopCount} distinct)`);

    if (distinctLoopCount > 0) {
        const [first] = loops.map(l => l.pathString).sort();

        console.log(`    affected Packages: [${[...loopPathMap.keys()].join(", ")}]`);
        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(groupedLicenses: GroupedLicenseSummary, paddingLeft: number): void {
    console.log(`Licenses:`);
    for (const { license, names } of groupedLicenses) {
        const threshold = 4;
        const samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${
                    names.length - threshold
                } more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }
}
