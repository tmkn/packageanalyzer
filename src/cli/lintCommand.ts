import { Command, Option } from "clipanion";

import { type ILintServiceConfig, LintService } from "../reports/lint/LintService.js";
import { LintFileLoader } from "../reports/lint/RulesLoader.js";
import { getPackageVersionfromString } from "../visitors/visitor.js";
import { npmOnline } from "../providers/online.js";
import { FileSystemPackageProvider } from "../providers/folder.js";
import { Formatter, type IFormatter } from "../utils/formatter.js";
import { getPackageVersionFromPath } from "../visitors/util.node.js";

export class LintCommand extends Command {
    public lintFile = Option.String();

    public depth = Option.String(`--depth`, "no_valid_number", {
        description: `how deep to analyze the dependency tree`
    });

    public package = Option.String(`--package`, {
        description: `the package to analyze e.g. typescript, typescript@3.5.1`
    });

    public folder = Option.String(`--folder`, { description: `path to a package.json` });

    static override usage = Command.Usage({
        description: `lint your dependency tree`,
        details: `
            This command will lint your dependency tree and print out the results
        `,
        examples: [
            [
                `Lint a local project`,
                `$0 lint --folder /path/to/your/package.json /path/to/lintconfig.js`
            ],
            [`Lint a dependency`, `$0 lint --package package /path/to/lintconfig.js`],
            [
                `Lint a specific version of a dependency`,
                `$0 lint --package package@version /path/to/lintconfig.js`
            ]
        ]
    });

    static override paths = [[`lint`]];

    async execute() {
        try {
            const config = this.#getConfig();
            this.beforeProcess?.(config);
            const lintService = new LintService(config, this.context.stdout, this.context.stderr);

            return lintService.process();
        } catch (e: any) {
            const stderrFormatter: IFormatter = new Formatter(this.context.stderr);

            stderrFormatter.writeLine(e?.toString());
            console.error(e?.toString());

            return 1;
        }
    }

    #getConfig(): ILintServiceConfig {
        let depth = parseInt(this.depth, 10);

        if (isNaN(depth)) {
            depth = Infinity;
        }

        if (this.package) {
            return {
                entry: getPackageVersionfromString(this.package),
                loader: new LintFileLoader(this.lintFile),
                depth,
                provider: npmOnline
            };
        } else if (this.folder) {
            return {
                entry: getPackageVersionFromPath(this.folder),
                loader: new LintFileLoader(this.lintFile),
                depth,
                provider: new FileSystemPackageProvider(this.folder)
            };
        } else {
            throw new Error(`No package nor folder option was provided`);
        }
    }

    beforeProcess: ((config: ILintServiceConfig) => void) | undefined = undefined;
}
