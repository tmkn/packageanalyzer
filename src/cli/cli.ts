import { Cli, Builtins } from "clipanion";

import { AnalyzeCommand } from "./analyzeCommand";
import { UpdateInfoCommand } from "./updateInfoCommand";
import { DownloadCommand } from "./downloadCommand";
import { LoopsCommand } from "./loopsCommand";
import { TreeCommand } from "./treeCommand";
import { NpmDumpCommand } from "./npmDumpCommand";
import { getVersion } from "./common";
import { NpmDumpLookupCreatorCommand } from "./npmLookupCreatorCommand";
import { LicenseCheckCommand } from "./licenseCommand";
import { ReportCommand } from "./reportCommand";
import { DependencyDumperCommand } from "./dependencyDumpCommand";

export const cli = new Cli({
    binaryLabel: `packageanalyzer`,
    binaryName: `pkga`,
    binaryVersion: getVersion()
});

//standard commands
cli.register(ReportCommand);
cli.register(AnalyzeCommand);
cli.register(UpdateInfoCommand);
cli.register(DownloadCommand);
cli.register(LoopsCommand);
cli.register(TreeCommand);
cli.register(LicenseCheckCommand);

//development niche commands
cli.register(NpmDumpCommand);
cli.register(NpmDumpLookupCreatorCommand);
cli.register(DependencyDumperCommand);

//built in commands
cli.register(Builtins.HelpCommand);
cli.register(Builtins.VersionCommand);