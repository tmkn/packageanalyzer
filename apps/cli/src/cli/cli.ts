import { Cli, Builtins } from "clipanion";

import { AnalyzeCommand } from "./analyzeCommand.js";
import { UpdateInfoCommand } from "./updateInfoCommand.js";
import { DownloadCommand } from "./downloadCommand.js";
import { LoopsCommand } from "./loopsCommand.js";
import { TreeCommand } from "./treeCommand.js";
import { LicenseCheckCommand } from "./licenseCommand.js";
import { ReportCommand } from "./reportCommand.js";
import { DependencyDumperCommand } from "./dependencyDumpCommand.js";
import { TestCommand } from "./testCommand.js";
import { DiffCommand } from "./diffCommand.js";
import { LintCommand } from "./lintCommand.js";
import { getVersion } from "../../../../packages/node/src/common.js";

export const cli = new Cli({
    binaryLabel: `packageanalyzer`,
    binaryName: `pkga`,
    binaryVersion: getVersion()
});

//standard commands
cli.register(ReportCommand);
cli.register(AnalyzeCommand);
cli.register(UpdateInfoCommand);
cli.register(DownloadCommand);
cli.register(LoopsCommand);
cli.register(TreeCommand);
cli.register(LicenseCheckCommand);
cli.register(DiffCommand);
cli.register(LintCommand);

//development niche commands
cli.register(DependencyDumperCommand);

cli.register(TestCommand);

//built in commands
cli.register(Builtins.HelpCommand);
cli.register(Builtins.VersionCommand);
