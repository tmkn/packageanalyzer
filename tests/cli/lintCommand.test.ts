import * as path from "path";
import { spawnSync } from "child_process";

import { cli } from "../../src/cli/cli.js";
import { LintCommand } from "../../src/cli/lintCommand.js";
import { createMockContext } from "../common.js";
import { type IMockPackageJson, MockProvider } from "../mocks.js";
import { pathToFileURL } from "url";

describe(`Lint Command`, () => {
    const medalloPkg: IMockPackageJson = {
        name: `medallo`,
        version: `1.0.0`,
        dependencies: [
            { name: `foo`, version: `1.0.0` },
            { name: `bar`, version: `1.0.0` }
        ]
    };
    const provider = new MockProvider([medalloPkg]);

    describe(`Error Handling`, () => {
        test(`fails on missing lint file`, async () => {
            const command = cli.process([
                `lint`,
                `--package`,
                `foo`,
                `./tests/mocks/lint/missing.js`
            ]) as LintCommand;

            expect.assertions(3);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;

            await command.execute();

            expect(stdout.lines.length).toBe(0);
            expect(stderr.lines.length).toBe(2);
            expect(stderr.lines[0]).toContain(`Error:`);
        });

        test(`fails on invalid lint file format`, async () => {
            const command = cli.process([
                `lint`,
                `--package`,
                `foo`,
                path.join(process.cwd(), `tests`, `sampleReport.js`)
            ]) as LintCommand;

            expect.assertions(3);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;

            await command.execute();

            expect(stdout.lines.length).toBe(0);
            expect(stderr.lines.length).toBe(2);
            expect(stderr.lines[0]).toContain(`Error:`);
        });

        test(`fails on missing --package and --folder`, async () => {
            const command = cli.process([
                `lint`,
                path.join(process.cwd(), `tests`, `sampleReport.js`)
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });

        test(`shows error message for invalid check return value`, async () => {
            const packageJsonPath = path.join(process.cwd(), "tests", "data", "lint_data");
            const lintFilePath = path.join(
                process.cwd(),
                "tests",
                "data",
                "lint_data",
                "lintFileInvalidCheck.js"
            );

            const command = cli.process([
                `lint`,
                `--folder`,
                packageJsonPath,
                lintFilePath
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });
    });

    describe(`--package`, () => {
        test(`correctly shows success message`, async () => {
            const command = cli.process([
                `lint`,
                `--package`,
                `medallo@1.0.0`,
                path.join(process.cwd(), `tests`, `sampleLintFileEmpty.js`)
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;
            command.beforeProcess = report => (report.provider = provider);

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });

        test(`correctly reports all packages in the tree`, async () => {
            const command = cli.process([
                `lint`,
                `--package`,
                `medallo@1.0.0`,
                path.join(process.cwd(), `tests`, `sampleLintFileNonEmpty.js`)
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;
            command.beforeProcess = report => (report.provider = provider);

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });
    });

    describe(`--folder`, () => {
        test(`correctly shows success message`, async () => {
            const packageJsonPath = path.join(process.cwd(), "tests", "data", "lint_data");
            const lintFilePath = path.join(
                process.cwd(),
                "tests",
                "data",
                "lint_data",
                "lintFile.js"
            );

            const command = cli.process([
                `lint`,
                `--folder`,
                packageJsonPath,
                lintFilePath
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });

        test(`correctly reports all packages as warnings`, async () => {
            const packageJsonPath = path.join(process.cwd(), "tests", "data", "lint_data");
            const lintFilePath = path.join(
                process.cwd(),
                "tests",
                "data",
                "lint_data",
                "allWarningLintFile.js"
            );

            const command = cli.process([
                `lint`,
                `--folder`,
                packageJsonPath,
                lintFilePath
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });

        test(`correctly reports all packages as errors`, async () => {
            const packageJsonPath = path.join(process.cwd(), "tests", "data", "lint_data");
            const lintFilePath = path.join(
                process.cwd(),
                "tests",
                "data",
                "lint_data",
                "allErrorLintFile.js"
            );

            const command = cli.process([
                `lint`,
                `--folder`,
                packageJsonPath,
                lintFilePath
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });
    });

    describe(`--depth`, () => {
        test(`--depth: 0  | correctly reports only root package in the tree`, async () => {
            const command = cli.process([
                `lint`,
                `--depth`,
                `0`,
                `--package`,
                `medallo@1.0.0`,
                path.join(process.cwd(), `tests`, `sampleLintFileNonEmpty.js`)
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;
            command.beforeProcess = report => (report.provider = provider);

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });

        test(`--depth: invalid  | correctly falls back to "Infinity" when depth is not valid`, async () => {
            const command = cli.process([
                `lint`,
                `--depth`,
                `invalid`,
                `--package`,
                `medallo@1.0.0`,
                path.join(process.cwd(), `tests`, `sampleLintFileNonEmpty.js`)
            ]) as LintCommand;

            expect.assertions(2);
            const { mockContext, stdout, stderr } = createMockContext();
            command.context = mockContext;
            command.beforeProcess = report => (report.provider = provider);

            await command.execute();

            expect(stdout.lines).toMatchSnapshot(`stdout`);
            expect(stderr.lines).toMatchSnapshot(`stderr`);
        });
    });

    describe(`exit codes`, () => {
        test(`returns 1 when reports are all 'error'`, async () => {
            const cwd = path.join(__dirname, "..", "..");
            const cliPath = path.join("build", "src", "cli.js");
            const packageJsonPath = path.join("tests", "data", "lint_data");
            const lintFilePath = pathToFileURL(
                path.join("tests", "data", "lint_data", "allErrorLintFile.js")
            ).href;
            const { status, stdout, stderr } = spawnSync(
                "node",
                [cliPath, "lint", "--folder", packageJsonPath, lintFilePath],
                {
                    encoding: "utf-8",
                    stdio: "pipe",
                    cwd
                }
            );

            expect(status).toBe(1);
        });

        test(`returns 1 when reports are mixed ('error' & 'warning')`, async () => {
            const cwd = path.join(__dirname, "..", "..");
            const cliPath = path.join("build", "src", "cli.js");
            const packageJsonPath = path.join("tests", "data", "lint_data");
            const lintFilePath = pathToFileURL(
                path.join("tests", "data", "lint_data", "allMixedLintFile.js")
            ).href;
            const { status, stdout, stderr } = spawnSync(
                "node",
                [cliPath, "lint", "--folder", packageJsonPath, lintFilePath],
                {
                    encoding: "utf-8",
                    stdio: "pipe",
                    cwd
                }
            );

            expect(status).toBe(1);
        });

        test(`returns 0 when reports are all 'warning'`, async () => {
            const cwd = path.join(__dirname, "..", "..");
            const cliPath = path.join("build", "src", "cli.js");
            const packageJsonPath = path.join(cwd, "tests", "data", "lint_data");
            const lintFilePath = pathToFileURL(
                path.join("tests", "data", "lint_data", "allWarningLintFile.js")
            ).href;
            const { status, stdout, stderr } = spawnSync(
                "node",
                [cliPath, "lint", "--folder", packageJsonPath, lintFilePath],
                {
                    encoding: "utf-8",
                    stdio: "pipe",
                    cwd
                }
            );

            expect(status).toBe(0);
        });
    });
});
