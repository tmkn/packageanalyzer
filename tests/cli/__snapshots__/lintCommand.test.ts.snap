// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lint Command --depth --depth: 0  | correctly reports only root package in the tree: stderr 1`] = `[]`;

exports[`Lint Command --depth --depth: 0  | correctly reports only root package in the tree: stdout 1`] = `
[
  "PackageLint: medallo@1.0.0",
  "",
  "medallo@1.0.0",
  "  [warning][medallo@1.0.0][sample-check]: This is a sample check for medallo",
  "",
  "Found 1 warning(s) and 0 error(s)",
  "",
  "",
]
`;

exports[`Lint Command --depth --depth: invalid  | correctly falls back to "Infinity" when depth is not valid: stderr 1`] = `[]`;

exports[`Lint Command --depth --depth: invalid  | correctly falls back to "Infinity" when depth is not valid: stdout 1`] = `
[
  "PackageLint: medallo@1.0.0",
  "",
  "medallo@1.0.0",
  "  [warning][medallo@1.0.0][sample-check]: This is a sample check for medallo",
  "",
  "medallo@1.0.0 → foo@1.0.0",
  "  [warning][foo@1.0.0][sample-check]: This is a sample check for foo",
  "",
  "medallo@1.0.0 → bar@1.0.0",
  "  [warning][bar@1.0.0][sample-check]: This is a sample check for bar",
  "",
  "Found 3 warning(s) and 0 error(s)",
  "",
  "",
]
`;

exports[`Lint Command --folder correctly reports all packages as errors: stderr 1`] = `[]`;

exports[`Lint Command --folder correctly reports all packages as errors: stdout 1`] = `
[
  "PackageLint: lint-test@1.0.0",
  "",
  "lint-test@1.0.0",
  "  [error][lint-test@1.0.0][sample-check]: This is a sample check for lint-test",
  "",
  "lint-test@1.0.0 → dependency1@1.0.0",
  "  [error][dependency1@1.0.0][sample-check]: This is a sample check for dependency1",
  "",
  "lint-test@1.0.0 → dependency2@2.0.0",
  "  [error][dependency2@2.0.0][sample-check]: This is a sample check for dependency2",
  "",
  "Found 0 warning(s) and 3 error(s)",
  "",
  "",
]
`;

exports[`Lint Command --folder correctly reports all packages as warnings: stderr 1`] = `[]`;

exports[`Lint Command --folder correctly reports all packages as warnings: stdout 1`] = `
[
  "PackageLint: lint-test@1.0.0",
  "",
  "lint-test@1.0.0",
  "  [warning][lint-test@1.0.0][sample-check]: This is a sample check for lint-test",
  "",
  "lint-test@1.0.0 → dependency1@1.0.0",
  "  [warning][dependency1@1.0.0][sample-check]: This is a sample check for dependency1",
  "",
  "lint-test@1.0.0 → dependency2@2.0.0",
  "  [warning][dependency2@2.0.0][sample-check]: This is a sample check for dependency2",
  "",
  "Found 3 warning(s) and 0 error(s)",
  "",
  "",
]
`;

exports[`Lint Command --folder correctly shows success message: stderr 1`] = `[]`;

exports[`Lint Command --folder correctly shows success message: stdout 1`] = `
[
  "PackageLint: lint-test@1.0.0",
  "",
  "Found no issues (0 warnings, 0 errors)",
  "",
  "",
]
`;

exports[`Lint Command --package correctly reports all packages in the tree: stderr 1`] = `[]`;

exports[`Lint Command --package correctly reports all packages in the tree: stdout 1`] = `
[
  "PackageLint: medallo@1.0.0",
  "",
  "medallo@1.0.0",
  "  [warning][medallo@1.0.0][sample-check]: This is a sample check for medallo",
  "",
  "medallo@1.0.0 → foo@1.0.0",
  "  [warning][foo@1.0.0][sample-check]: This is a sample check for foo",
  "",
  "medallo@1.0.0 → bar@1.0.0",
  "  [warning][bar@1.0.0][sample-check]: This is a sample check for bar",
  "",
  "Found 3 warning(s) and 0 error(s)",
  "",
  "",
]
`;

exports[`Lint Command --package correctly shows success message: stderr 1`] = `[]`;

exports[`Lint Command --package correctly shows success message: stdout 1`] = `
[
  "PackageLint: medallo@1.0.0",
  "",
  "Found no issues (0 warnings, 0 errors)",
  "",
  "",
]
`;

exports[`Lint Command Error Handling fails on missing --package and --folder: stderr 1`] = `
[
  "Error: No package nor folder option was provided",
  "",
]
`;

exports[`Lint Command Error Handling fails on missing --package and --folder: stdout 1`] = `[]`;

exports[`Lint Command Error Handling shows error message for invalid check return value: stderr 1`] = `[]`;

exports[`Lint Command Error Handling shows error message for invalid check return value: stdout 1`] = `
[
  "PackageLint: lint-test@1.0.0",
  "",
  "lint-test@1.0.0",
  "  [internal-error][lint-test@1.0.0][sample-check]: Invalid check implementation! check() must return "string" or "string[]". Returned "number"",
  "",
  "lint-test@1.0.0 → dependency1@1.0.0",
  "  [internal-error][dependency1@1.0.0][sample-check]: Invalid check implementation! check() must return "string" or "string[]". Returned "number"",
  "",
  "lint-test@1.0.0 → dependency2@2.0.0",
  "  [internal-error][dependency2@2.0.0][sample-check]: Invalid check implementation! check() must return "string" or "string[]". Returned "number"",
  "",
  "Found 0 warning(s) and 0 error(s)",
  "Terminated with 3 internal error(s), please check lint output",
  "",
  "",
]
`;
